src/puppet.js

```diff
@@ -89,7 +89,145 @@
 
   }
 
-  function PuppetNetworkChannel(puppet, remoteUrl, useWebSocket, onReceive, onSend, onError, onStateChange) {
+  /**
+   * @callback reconnectionCallback called when reconnection attempt is scheduled.
+   * It's called every second until reconnection attempt is made (`seconds` reaches 0)
+   * @param {number} seconds - number of seconds to next reconnection attempt. >= 0
+   */
+  /**
+   * @param {Function} reconnect used to perform reconnection. No arguments
+   * @param {reconnectionCallback} onReconnectionCountdown called to notify that reconnection attempt is scheduled
+   * @param {Function} onReconnectionEnd called to notify that reconnection attempt is not longer scheduled
+   * @constructor
+   */
+  function Reconnector(reconnect, onReconnectionCountdown, onReconnectionEnd) {
+    var intervalMs,
+        timeToCurrentReconnectionMs,
+        reconnectionPending,
+        reconnection,
+        defaultIntervalMs = 1000;
+
+    var reset = (function () {
```

Comment:
Is there any reason to keep the first parenthesis? 

Plus, I think as you are declaring them at the beginning, you can declare them as functions not function expressions.